File: packages/host-rest/src/routes/binding.ts
import { applyActionEffects, type ParsedInvokableRef, parseInvokableRef } from "@kohaku-ui/host-core";
import type { Principal, VerifyResult } from "@kohaku-ui/spec-core";
import type { Context, Hono } from "hono";
import { errorBody } from "../errors.js";
import type { KohakuHostDeps } from "../types.js";
import { ActionBodySchema } from "./schemas.js";
import { ANONYMOUS, message, parseBody, type RouteContext, reportHostError, requestIdOf } from "./shared.js";

/**
 * A raw downstream (DomainPort.invoke) failure never reaches the client verbatim on the REF_NOT_FOUND path: it
 * may carry internals (SQL fragments, stack-trace text, library-internal wording). Classifying a permanent
 * "no such reference" versus a transient failure would require a typed Port contract, so both collapse to this
 * fixed message; the original error still reaches the observability hook (onError) via reportHostError.
 */
const REF_NOT_FOUND_MESSAGE = "reference not found or not resolvable";

/**
 * Resolves the principal to act as for a verified capability. When `deps.auth` is wired (the host performs
 * real authentication), a `verify` that returns `ok:true` without a principal is a capability-issuer
 * misconfiguration and must not silently degrade to ANONYMOUS — doing so would let an authenticated
 * deployment act as the demo user. Denied with 403 CAPABILITY_DENIED in that case. When `deps.auth` is unwired
 * (the unauthenticated demo path), ANONYMOUS is the expected, intentional principal.
 * Returns the Response to return as-is on denial, or the resolved Principal on success.
 */
function resolvePrincipal(deps: KohakuHostDeps, c: Context, verdict: VerifyResult): Response | Principal {
  if (verdict.principal != null) return verdict.principal;
  if (deps.auth == null) return ANONYMOUS;
  return c.json(errorBody("CAPABILITY_DENIED", "capability verified without a principal"), 403);
}

/** Reference-passing data resolution / write-through path (/binding/resolve, /binding/action). capability required. */
export function registerBindingRoutes(app: Hono, ctx: RouteContext): void {
  const { deps } = ctx;

  // --- Reference-passing data resolution (component -> API direct. capability required) ---
  app.get("/binding/resolve", async (c) => {
    const refParam = c.req.query("ref");
    if (refParam == null) {
      return c.json(errorBody("BAD_REQUEST", "ref query parameter is required"), 400);
    }
    const token = bearerToken(c);
    if (token == null) {
      return c.json(errorBody("CAPABILITY_REQUIRED", "Authorization: Bearer <capability> is required"), 401);
    }

    // Server-side paging/sorting: parse ref into base (with reserved params removed) and reserved via
    // host-core's parseInvokableRef (shared with the MCP profile's resolve_binding tool / initial-data
    // preresolution). capability verification is an exact match against base.raw (= the canonical form of
    // the $ref the Spec declared). Only known reserved-param keys (_cursor/_limit/_sort/_dir) are allowed —
    // since the reserved namespace is outside authorization checks, passing an unknown `_` key through would
    // let the data range be changed with parameters outside the capability.
    let parsed: ParsedInvokableRef;
    try {
      parsed = parseInvokableRef(refParam, deps.querySource);
    } catch (e) {
      return c.json(errorBody("BAD_REQUEST", message(e)), 400);
    }
    if (parsed.kind === "source_mismatch") {
      return c.json(errorBody("SOURCE_MISMATCH", `unknown query source "${parsed.source}"`), 404);
    }
    const { base, params } = parsed.ref;

    const verdict = await deps.authz.verify(token, { kind: "read", ref: base.raw });
    if (!verdict.ok) {
      return c.json(errorBody("CAPABILITY_DENIED", verdict.reason ?? "capability denied"), 403);
    }
    const principal = resolvePrincipal(deps, c, verdict);
    if (principal instanceof Response) return principal;

    try {
      // Reserved params (_cursor/_limit/_sort/_dir) are already merged into params by parseInvokableRef
      // (the `_` namespace convention — DomainPort signature unchanged).
      const data = await deps.domain.invoke(base.path, params, { principal, capability: token });
      return c.json(data as object);
    } catch (e) {
      // Downstream (DomainPort) failures are also placed on the observability hook (symmetric with the other
      // composition handlers' failure-path observability), rather than turning into a 404 that goes unobserved.
      // The response stays REF_NOT_FOUND (404) — classifying
      // a permanent "no such reference" versus a transient failure entails introducing a Port contract (typed
      // errors), so that is left as a separate decision. The raw error message never reaches the client (it may
      // leak internals); the original error still reaches onError via reportHostError.
      const requestId = requestIdOf(c, deps);
      await reportHostError(deps, "binding/resolve", requestId, e);
      return c.json(errorBody("REF_NOT_FOUND", REF_NOT_FOUND_MESSAGE, requestId), 404);
    }
  });

  // --- Write-through path (presentForm submit, etc.) ---
  app.post("/binding/action", async (c) => {
    const body = await parseBody(c, ActionBodySchema, "action is required");
    if (body instanceof Response) return body;
    const token = bearerToken(c);
    if (token == null) {
      return c.json(errorBody("CAPABILITY_REQUIRED", "Authorization: Bearer <capability> is required"), 401);
    }
    const verdict = await deps.authz.verify(token, { kind: "write", ref: body.action });
    if (!verdict.ok) {
      return c.json(errorBody("CAPABILITY_DENIED", verdict.reason ?? "capability denied"), 403);
    }
    const principal = resolvePrincipal(deps, c, verdict);
    if (principal instanceof Response) return principal;
    // body.payload is already a validated JsonObject | undefined (ActionBodySchema); no cast needed.
    const payload = body.payload ?? {};
    let result: unknown;
    try {
      result = await deps.domain.invoke(body.action, payload, { principal, capability: token });
    } catch (e) {
      // Failure of the write itself (domain.invoke) is 404. Place the downstream failure on the
      // observability hook, then map it. The raw error message never reaches the client (see REF_NOT_FOUND_MESSAGE).
      const requestId = requestIdOf(c, deps);
      await reportHostError(deps, "binding/action", requestId, e);
      return c.json(errorBody("REF_NOT_FOUND", REF_NOT_FOUND_MESSAGE, requestId), 404);
    }
    // Write-already-committed vs. side-effect-declaration failure: see host-core's applyActionEffects.
    const response = await applyActionEffects(deps.actionEffects, body.action, payload, result, async (e) => {
      const requestId = requestIdOf(c, deps);
      await reportHostError(deps, "binding/action.effects", requestId, e);
    });
    return c.json(response);
  });
}

function bearerToken(c: Context): string | null {
  const header = c.req.header("authorization");
  if (header == null || !header.toLowerCase().startsWith("bearer ")) return null;
  return header.slice(7).trim();
}

File: python/kohaku/src/kohaku/host_rest/_routes/binding.py
"""Binding route group: /binding/resolve (reference-passing data resolution) and /binding/action (direct write path).

Split out of the former monolithic `_fastapi_routes.py` to mirror packages/host-rest/src/routes/binding.ts.
"""

from __future__ import annotations

from fastapi import APIRouter
from starlette.requests import Request
from starlette.responses import Response

from kohaku.host_core import ParsedInvokableRefOk, apply_action_effects, parse_invokable_ref
from kohaku.spec import InvocationContext, Principal, QueryRefError, VerifyRequest, VerifyResult

from ..bodies import parse_action_body
from ..deps import KohakuHostDeps
from .shared import (
    ANONYMOUS,
    _bearer_token,
    _error,
    _json,
    _message,
    _read_json,
    report_host_error,
    request_id_of,
)

# A raw downstream (DomainPort.invoke) failure never reaches the client verbatim on the REF_NOT_FOUND path: it
# may carry internals (SQL fragments, stack-trace text, library-internal wording). The original error still
# reaches the observability hook (on_error) via report_host_error.
_REF_NOT_FOUND_MESSAGE = "reference not found or not resolvable"


def _resolve_principal(deps: KohakuHostDeps, verdict: VerifyResult) -> Principal | Response:
    """Resolve the principal to act as for a verified capability. When deps.auth is wired (the host performs
    real authentication), a verify that returns ok=True without a principal is a capability-issuer
    misconfiguration and must not silently degrade to ANONYMOUS — doing so would let an authenticated
    deployment act as the demo user. Denied with 403 CAPABILITY_DENIED in that case. When deps.auth is unwired
    (the unauthenticated demo path), ANONYMOUS is the expected, intentional principal.
    Returns the Response to return as-is on denial, or the resolved Principal on success.
    """
    if verdict.principal is not None:
        return verdict.principal
    if deps.auth is None:
        return ANONYMOUS
    return _error("CAPABILITY_DENIED", "capability verified without a principal", 403)


def register_binding_routes(router: APIRouter, deps: KohakuHostDeps) -> None:
    """Registers /binding/resolve and /binding/action onto router."""

    # --- Reference-passing data resolution -----------------------------------
    @router.get("/binding/resolve")
    async def binding_resolve(request: Request) -> Response:
        ref_param = request.query_params.get("ref")
        if ref_param is None:
            return _error("BAD_REQUEST", "ref query parameter is required", 400)
        token = _bearer_token(request)
        if token is None:
            return _error(
                "CAPABILITY_REQUIRED", "Authorization: Bearer <capability> is required", 401
            )
        # Server-side paging/sorting: parse ref into base (with reserved params removed) and reserved via
        # host_core's parse_invokable_ref (shared with the MCP profile's resolve_binding tool / initial-data
        # preresolution). Capability verification is an exact match against base.raw (= the canonical form of
        # the $ref the Spec declared). Only known reserved-param keys (_cursor/_limit/_sort/_dir) are allowed —
        # since the reserved namespace is outside authorization checks, passing an unknown `_` key through would
        # let the data range be changed with parameters outside the capability.
        try:
            parsed = parse_invokable_ref(ref_param, deps.query_source)
        except (QueryRefError, ValueError) as e:
            return _error("BAD_REQUEST", _message(e), 400)
        if not isinstance(parsed, ParsedInvokableRefOk):
            return _error("SOURCE_MISMATCH", f'unknown query source "{parsed.source}"', 404)
        base, params = parsed.ref.base, parsed.ref.params
        verdict = await deps.authz.verify(token, VerifyRequest(kind="read", ref=base.raw))
        if not verdict.ok:
            return _error("CAPABILITY_DENIED", verdict.reason or "capability denied", 403)
        principal = _resolve_principal(deps, verdict)
        if isinstance(principal, Response):
            return principal
        try:
            data = await deps.domain.invoke(
                base.path,
                params,
                InvocationContext(principal=principal, capability=token),
            )
            return _json(data)
        except BaseException as e:
            # The raw error message never reaches the client (it may leak internals); the original error still
            # reaches on_error via report_host_error.
            request_id = request_id_of(request, deps)
            await report_host_error(deps, "binding/resolve", request_id, e)
            return _error("REF_NOT_FOUND", _REF_NOT_FOUND_MESSAGE, 404, request_id)

    # --- Direct write path ----------------------------------------------------
    @router.post("/binding/action")
    async def binding_action(request: Request) -> Response:
        body = parse_action_body(await _read_json(request))
        if body is None:
            return _error("BAD_REQUEST", "action is required", 400)
        token = _bearer_token(request)
        if token is None:
            return _error(
                "CAPABILITY_REQUIRED", "Authorization: Bearer <capability> is required", 401
            )
        verdict = await deps.authz.verify(token, VerifyRequest(kind="write", ref=body.action))
        if not verdict.ok:
            return _error("CAPABILITY_DENIED", verdict.reason or "capability denied", 403)
        principal = _resolve_principal(deps, verdict)
        if isinstance(principal, Response):
            return principal
        payload = body.payload if body.payload is not None else {}
        try:
            result = await deps.domain.invoke(
                body.action,
                payload,
                InvocationContext(principal=principal, capability=token),
            )
        except BaseException as e:
            # A failure of the write itself (domain.invoke) is not yet committed, so respond with an error.
            # The raw error message never reaches the client (see _REF_NOT_FOUND_MESSAGE).
            request_id = request_id_of(request, deps)
            await report_host_error(deps, "binding/action", request_id, e)
            return _error("REF_NOT_FOUND", _REF_NOT_FOUND_MESSAGE, 404, request_id)
        # Write-already-committed vs. side-effect-declaration failure: see host_core's apply_action_effects.
        # wide_catch=True preserves REST's pre-branch `except BaseException` at this call site.
        response = await apply_action_effects(
            deps.action_effects,
            body.action,
            payload,
            result,
            lambda e: report_host_error(deps, "binding/action", request_id_of(request, deps), e),
            wide_catch=True,
        )
        return _json(response)
